_posts/2018-03-15-complete-introduction-to-nodejs.md

```diff
@@ -0,0 +1,406 @@
+  req.on('readable', ()=>{
+      let chunk = null;
+      // as long as we can read chunks from the request, we write those chunks to the response.
+      while(null !== (chunk = request.read())){
+         // we can keep writing to the writable response stream as long as the connection is open,
+         // so we keep piping the readable data to the response
+          response.write(chunk);
+      }
+      // you can test it with curl from the terminal by sending a 'hello' string as data,
+      // simply run: curl -d 'hello' http://localhost:3000
+      // the 'hello' string is being sent back to the client
+  });
+
+  req.on('end', ()=>{
+      // once the request stream is closed, we also close the response stream
+      response.end();
+  });
+
+  /*
+	I've written it out completely to show you what is does,
+	but in fact we can use the pipe method to refactor the request's 'on readable' in some much simpler code
+	req.on('readable', ()=>{
+             request.pipe(response)
+        })
+   */
+});
+
+server.listen(3000, ()=>{
+    console.log('Server running at localhost:3000/');
+});
+```
+
+## Cluster
+A single instance of Node.js runs in a single thread.
+To take advantage of multi-core systems, the user will sometimes want to launch a cluster of Node.js processes to handle the load.
+
+```javascript
+// require the cluster module
+const cluster = require('cluster');
+// we'll set up a http server on all cpu's and load balance between them
+const http = require('http');
+// we need to know the amount of cpu's our machine has available,
+// so we do this with the 'os' module's cpus method which returns an array of cpus, to get the amount we get the array's length
+const numCPUs = require('os').cpus().length;
+
+if (cluster.isMaster) {
+  // the cluster will first start up a master process that forks itself onto the other cpu's and handles the load balancing between these workers
+  console.log(`Master ${process.pid} is running`);
+
+  // fork this process to a worker for every cpu that is left (not the < and not <=)
+  for (let i = 0; i < numCPUs; i++) {
+    cluster.fork();
+  }
+
+  // when a worker dies, it emits an exit event
+  cluster.on('exit', (worker, code, signal) => {
+    console.log(`worker ${worker.process.pid} died`);
+  });
+} else {
+  // if the cluster is not master, it (in this case) sets up our http server,
+  // Workers can share any TCP connection
+
+  http.createServer((req, res) => {
+    res.writeHead(200);
+    res.end('hello world\n');
+  }).listen(8000);
+
+  console.log(`Worker ${process.pid} started`);
+}
+
+```
+More information on clusters on [https://nodejs.org/api/cluster.html](https://nodejs.org/api/cluster.html)
+
+
+Finally
+So that's it for this blogpost. I hope it will be useful for you.
+If you have any suggestions or feel like I've forgotten to mention some important stuff, feel free to comment below.
+I'm currently working on some follow up tutorials:
```

Comment:
follow-up

Reply:
fixed in https://github.com/ordina-jworks/ordina-jworks.github.io/pull/359/commits/daf5636773b89b76532c425468fddc6511bf0643